dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/ForcedMaxOutputRowsRule.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.sql.watchdog
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.plans.logical.{GlobalLimit, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.Origin
+import org.apache.spark.sql.execution.command.InsertIntoDataSourceDirCommand
+
+import org.apache.kyuubi.sql.KyuubiSQLConf
+
+/*
+* Add ForcedMaxOutputRows rule for output rows limitation to avoid huge output rows of non_limit query unexpectedly
+* mainly applied to cases as below:
+* case 1:
+*  {{{
+*   INSERT OVERWRITE DIRECTORY (path=STRING)?
+*   USING format OPTIONS ([option1_name "option1_value", option2_name "option2_value", ...])
+*   SELECT ...
+* }}}
+*
+* case 2:
+* {{{
+*   SELECT [c1, c2, ...]
+* }}}
+*
+* case 3:
+* {{{
+*   WITH CTE AS (
+*   ...)
+* SELECT [c1, c2, ...] FROM CTE ...
+* }}}
+*
+* The Logical Rule add a GlobalLimit node before root project
+* */
+case class ForcedMaxOutputRowsRule(session: SparkSession) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!plan.resolved) {
+      plan
+    } else {
+      conf.getConf(KyuubiSQLConf.WATCHDOG_FORCED_MAXOUTPUTROWS) match {
+        case Some(forcedMaxOutputRows) => plan match {
+          case insert@InsertIntoDataSourceDirCommand(_, _, Project(_, _), _) =>
+            insert.copy(query = GlobalLimit(forcedMaxOutputRows, insert.query))
+          case project@Project(_, _) => project.origin match {
+            case Origin(None, None) => GlobalLimit(forcedMaxOutputRows, project)
+            case Origin(Some(_), Some(0)) => GlobalLimit(forcedMaxOutputRows, project)
```

Comment:
why we need match `Origin` if the plan is resolved ?

Reply:
> why we need match `Origin` if the plan is resolved ?

1. It can actually remove `case Origin(None, None) => GlobalLimit(forcedMaxOutputRows, project)` before plan is resolved
2. Match `Origin` mainly applied to `SELECT ...` or `WITH CTE(...) SELECT...`, output nodes such as `Project` always starts  0 position on Analyzer，so it judge the output nodes depend on `Origin(_, Some(0))`

Reply:
what's difference if we only match `Project` ? do you mean we will get a wrong node to add limit ?

Reply:
SQL trees may contain other node of same type, especially there may have `Project` or `Aggregate`  in sub-query
It's critical to find the root output node such as `Project` or `Aggregate`, so root `Project match Origin(Some(_), Some(0)` and `Aggregate match Origin(None, None)`

Reply:
Seems a little hack, the `Origin` is used to distinguish if the plan is subquery ?

Reply:
indeed, this implement is obscure, Shall we have other solution to find tree output root node, I want to know better way to instead this obscure match

Reply:
is there any issue if we also add limit for suquery ? I think we don't need care subquery as following:

- scalar subquery; it return at most one row or exception
- in subquery; it's ok to add limit so we can protect engine's driver (OOM)
- exist subquery; it will be rewrite to scalar subquery, so it's also safe.

Reply:
> Seems a little hack, the `Origin` is used to distinguish if the plan is subquery ?

The `Origin` provides a location for TreeNodes in the context of their origin, it define as below(reference: https://github.com/apache/spark/blob/branch-3.1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala#L49
):
```scala 
case class Origin(
  line: Option[Int] = None,
  startPosition: Option[Int] = None)
The chief output node as root with init position(None or Some(0)), it's not used to distinguish subquery, we can use subquery class to identify



> is there any issue if we also add limit for suquery ? I think we don't need care subquery as following:
> 
> * scalar subquery; it return at most one row or exception
> * in subquery; it's ok to add limit so we can protect engine's driver (OOM)
> * exist subquery; it will be rewrite to scalar subquery, so it's also safe.

Add limit for subquery may make difference with origin query，my implement will not limit node to subquery
Do you tend to add subquery with above three reasons or not？

Reply:
> is there any issue if we also add limit for suquery ? I think we don't need care subquery as following:
> 
> * scalar subquery; it return at most one row or exception
> * in subquery; it's ok to add limit so we can protect engine's driver (OOM)
> * exist subquery; it will be rewrite to scalar subquery, so it's also safe.

Add limit for subquery may make difference with origin query，my implement will not limit node to subquery
Do you tend to add subquery with above three reasons or not？

Reply:
> The Origin provides a location for TreeNodes in the context of their origin
> it's not used to distinguish subquery, we can use subquery class to identify

So I'm still not see why you need match origin here, the origin is used to help track error sql string. Is there any use case that Spark match the origin to do the simialr thing ?

Reply:
> > The Origin provides a location for TreeNodes in the context of their origin
> > it's not used to distinguish subquery, we can use subquery class to identify
> 
> So I'm still not see why you need match origin here, the origin is used to help track error sql string. Is there any use case that Spark match the origin to do the simialr thing ?

Why I use origin to judge the output root node?
According SQL parse principle, it parse with top-down, the output root node is the top node, so I depend on `startPosition` to find root node, It works on sql parse with `antlr` , now the implement mainly use the `startPosition ` in `Token` parsed by antlr, `startPosition` notes that The index of the first character of this token relative to the beginning of the line at which it occurs, 0..n-1

As you said `Origin` is used to track sql string, it almost not match `Origin` in spark

Reply:
does the project is not root node as following ?
```
plan match {
  case _: Project
}
```

can you give a case that we must match origin ?

Reply:
I think out this issue, the plan run `match` operator on this rule , it's just the root node, so it doesn't travel matched plan children, thank your heuristic!